### PR TITLE
trying full build

### PR DESCRIPTION
**Pull request by sarthakpati**
Checklist
* [ ] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by jaimergp** [comment]
This is failing with

```
OSError: dlopen($PREFIX/lib/python3.6/site-packages/torch_scatter/_scatter_cpu.so, 6): Symbol not found: __ZN2at8GradMode10is_enabledEv
  Referenced from: $PREFIX/lib/python3.6/site-packages/torch_scatter/_scatter_cpu.so
  Expected in: $PREFIX/lib/python3.6/site-packages/torch/lib/libtorch_cpu.dylib
 in $PREFIX/lib/python3.6/site-packages/torch_scatter/_scatter_cpu.so
```

That's a missing symbol in `libtorch_cpu`. Do you know if we need to pin against a specific `pytorch` version?

**Comment by sarthakpati** [comment]
I am not sure, perhaps @rusty1s could comment?

**Comment by rusty1s** [comment]
It looks like you are building with PyTorch 1.8.0 while testing with PyTorch 1.9.0.

**Comment by sarthakpati** [comment]
That's weird, because I have the same version requirements for all stages in the recipe [[ref](https://github.com/sarthakpati/torch-scatter-feedstock/blob/full_build/recipe/meta.yaml#L26)]: `pytorch >=1.7`

I thought hard-coding a specific version would be problematic?

File: recipe/meta.yaml [hunk 1]
```diff
@@ -32,6 +31,17 @@ requirements:
...
+    - scikit-learn
+    - requests
+    - pandas
+    - rdflib
+    - googledrivedownloader
+    - jinja2
+    - pyparsing
   run:
     - python
     - pytorch >=1.7
```
Comment (jaimergp) [inline recipe/meta.yaml:47]: Try this:

```suggestion
    - {{ pin_compatible('pytorch', max_pin='x.x.x') }}
```

File: recipe/meta.yaml [hunk 2]
```diff
@@ -32,6 +31,17 @@ requirements:
     - pytest-runner
     - python
     - pytorch >=1.7
```
Comment (jaimergp) [inline recipe/meta.yaml:33]: I think PyTorch is globally pinned as of 1.9.0. Is that right @hmaarrfk?

```suggestion
    - pytorch
```

File: recipe/meta.yaml (hunk 1 above)
Comment (jaimergp) [reply recipe/meta.yaml:47]: This shouldn't be needed as of 1.9.0, though :/

File: recipe/meta.yaml (hunk 2 above)
Comment (hmaarrfk) [reply recipe/meta.yaml:33]: pytorch as a build time dependency should be left to conda-forge-pinnings.

This suggestion is the way to go (I'm a maintainer in the pytorch recipe).

File: recipe/meta.yaml (hunk 1 above)
Comment (hmaarrfk) [reply recipe/meta.yaml:47]: This is correct, you should likely just leave it blank.

Comment (sarthakpati) [reply recipe/meta.yaml]: Is this fine? https://github.com/sarthakpati/torch-scatter-feedstock/blob/full_build/recipe/meta.yaml#L47

Comment (jaimergp) [reply recipe/meta.yaml]: Yep, that looks fine. Let's see how it goes!

Comment (jaimergp) [reply recipe/meta.yaml]: Thanks @hmaarrfk!

**Comment by sarthakpati** [comment]
@jaimergp @hmaarrfk: still failed. and for some reason, Linux builds are not getting initialized, either, even though I removed the "skip" requirement.

**Comment by jaimergp** [comment]
@conda-forge-admin, please rerender

**Comment by sarthakpati** [comment]
Looks much better. What do you guys think?

**Comment by hmaarrfk** [comment]
does this support GPU? if so. take a look at torchvision and copy over the cuda stuff.

**Comment by sarthakpati** [comment]
Thanks!

> does this support GPU?
Yes, it does. 
> if so. take a look at torchvision and copy over the cuda stuff.
How does this look?

**Comment by sarthakpati** [comment]
@conda-forge-admin, please rerender

**Comment by sarthakpati** [comment]
It is weird that the build process is unable to pick out the build scripts:

```
/home/conda/feedstock_root/build_artifacts/torch-scatter_1630691264698/work/conda_build.sh: line 4: build-torch.sh: command not found
```

Am I missing something, here? Those files are already present...

**Comment by hmaarrfk** [comment]
why not rename them to `build.sh` and `bld.bat` the other files are more for multiple outputs.

**Comment by sarthakpati** [comment]
> why not rename them to `build.sh` and `bld.bat` the other files are more for multiple outputs.

Okay, trying that now! Thank you.

File: recipe/meta.yaml [hunk 3]
```diff
@@ -11,30 +11,49 @@ source:
...
-  skip: true  # [win or (linux and cuda_compiler_version in (undefined, 'None', '10.2'))]
+  number: 1
   rpaths:
     - lib/
     - {{ SP_DIR }}/torch/lib
   missing_dso_whitelist:
     - '*/libcuda.*'  # [linux64 or ppc64le]
-  script: {{ PYTHON }} -m pip install . -vv
+  script: build.sh   # [not win]
+  script: bld.bat  # [win]
```
Comment (hmaarrfk) [inline recipe/meta.yaml:21]:
```suggestion
```

**Comment by sarthakpati** [comment]
Thanks a ton for your help @hmaarrfk and @jaimergp! 

I am getting this error:

```
/usr/local/cuda/include/crt/host_config.h:138:2: error: #error -- unsupported GNU version! gcc versions later than 8 are not supported!
    138 | #error -- unsupported GNU version! gcc versions later than 8 are not supported!
```

How can I ensure that the gcc version is acceptable for nvcc?

**Comment by hmaarrfk** [comment]
do you personally need cuda 10.2? if not, i would just skip it. there is a good chance it will be gone in a few months from the comda-forge build matrix

**Comment by hmaarrfk** [comment]
@conda-forge-admin please rerender

**Comment by sarthakpati** [comment]
> do you personally need cuda 10.2? if not, i would just skip it. there is a good chance it will be gone in a few months from the comda-forge build matrix

Nope, I do not.

**Comment by hmaarrfk** [comment]
you can have many skip like selectors. you can have one that checks for cuda 10.2

**Comment by sarthakpati** [comment]
Do you mean the checks for `# [cuda_compiler_version != "None"]`?

**Comment by sarthakpati** [comment]
@conda-forge-admin please rerender

**Comment by hmaarrfk** [comment]
image
Your selector didn't work out as planned.

**Comment by hmaarrfk** [comment]
Try adding the migration files from
https://github.com/conda-forge/torchvision-feedstock/tree/master/.ci_support/migrations

**Comment by sarthakpati** [comment]
@conda-forge-admin please rerender

**Comment by hmaarrfk** [comment]
Cool! And it seems that this recipe was previously skipping cuda 10.2 so I guess nothing is lost if this works.

Good luck!

**Comment by hmaarrfk** [comment]
usually you would let the migrator do its things, but it feel appropriate to test CUDA here since it somehow became part of the scope of this PR.

**Comment by sarthakpati** [comment]
Cool, thanks. Seems like this is now good to merge?

**Comment by hmaarrfk** [comment]
Yes. The CUDA builds seem to have pulled in the correct dependencies:

For example, cuda 11 pulled in
```
The following NEW packages will be INSTALLED:

    _libgcc_mutex:     0.1-conda_forge             conda-forge
    _openmp_mutex:     4.5-1_llvm                  conda-forge
    ca-certificates:   2021.5.30-ha878542_0        conda-forge
    cffi:              1.14.6-py36hc120d54_0       conda-forge
    cudatoolkit:       11.1.1-h6406543_8           conda-forge
    cudnn:             8.2.1.32-h86fa8c9_0         conda-forge
    dataclasses:       0.8-pyh787bdff_2            conda-forge
    future:            0.18.2-py36h5fab9bb_3       conda-forge
    ld_impl_linux-64:  2.36.1-hea4e1c9_2           conda-forge
    libblas:           3.9.0-11_linux64_mkl        conda-forge
    libcblas:          3.9.0-11_linux64_mkl        conda-forge
    libffi:            3.3-h58526e2_2              conda-forge
    libgcc-ng:         11.1.0-hc902ee8_8           conda-forge
    liblapack:         3.9.0-11_linux64_mkl        conda-forge
    libprotobuf:       3.16.0-h780b84a_0           conda-forge
    libstdcxx-ng:      11.1.0-h56837e0_8           conda-forge
    llvm-openmp:       12.0.1-h4bd325d_1           conda-forge
    magma:             2.5.4-ha9b7cf9_2            conda-forge
    mkl:               2021.3.0-h726a3e6_557       conda-forge
    nccl:              2.10.3.1-h97a9cb7_0         conda-forge
    ncurses:           6.2-h58526e2_4              conda-forge
    ninja:             1.10.2-h4bd325d_0           conda-forge
    numpy:             1.19.5-py36hfc0c790_2       conda-forge
    openssl:           1.1.1l-h7f98852_0           conda-forge
    pip:               21.2.4-pyhd8ed1ab_0         conda-forge
    pycparser:         2.20-pyh9f0ad1d_2           conda-forge
    python:            3.6.13-hffdb5ce_0_cpython   conda-forge
    python_abi:        3.6-2_cp36m                 conda-forge
    pytorch:           1.9.0-cuda111py36hc5445e6_1 conda-forge
    pytorch-gpu:       1.9.0-cuda111py36h788eb59_1 conda-forge
    readline:          8.1-h46c0cb4_0              conda-forge
    setuptools:        57.4.0-py36h5fab9bb_0       conda-forge
    sleef:             3.5.1-h7f98852_1            conda-forge
    typing_extensions: 3.10.0.0-pyha770c72_0       conda-forge
    wheel:             0.37.0-pyhd8ed1ab_1         conda-forge
    xz:                5.2.5-h516909a_1            conda-forge
    zlib:              1.2.11-h516909a_1010        conda-forge
```

Cuda compiler 11.0 with python 3.8 generated the build:
```
torch-scatter-2.0.8-py38h65b7e2b_1.tar.bz2
```

Cuda compiler 11.1 with python 3.8 generated the buidl:
```
torch-scatter-2.0.8-py38h7b4eac4_1.tar.bz2
```

So the build hashes are different. I think you are safe against collisions. 

I looks good to me.

**Comment by sarthakpati** [comment]
Awesome, thank you so much!

**Pull request merged**